File: 0494-target-sum/0494-target-sum.cpp
class Solution {
public:
    int total = 0;
    int findTargetSumWays(vector<int>& nums, int target) {
        int n = nums.size();
        for(int x:nums) {
            total += abs(x);
        }
        cout<<2*total + 1<<endl;
        vector<vector<int>>dp(n+1, vector<int>(2*total + 1, -1));
        return solve(nums, target, 0, dp, 0);
    }
    int solve(vector<int>& nums, int target, int cur, vector<vector<int>>&dp, int sum) {
        if(cur == nums.size()) {
            return target == sum;
        }
        if(dp[cur][sum + total] != -1) return dp[cur][sum + total];
        int ways1 = solve(nums, target, cur + 1, dp, sum + nums[cur]);
        int ways2 = solve(nums, target, cur + 1, dp, sum - nums[cur]);
        return dp[cur][sum + total] = ways1 + ways2;
    }
};

File: 0494-target-sum/0494-target-sum.java
class Solution {
    int total = 0;
    public int findTargetSumWays(int[] nums, int target) {
        int n = nums.length;
        for(int x:nums) {
            total += Math.abs(x);
        }
        int[][] dp = new int[n+1][2*total + 1];
        for(int i=0; i<=n; i++) {
            Arrays.fill(dp[i], -1);
        }
        return solve(nums, target, 0, 0, dp);
    }
    private int solve(int[] nums, int target, int cur, int sum, int[][] dp) {
        if(cur >= nums.length) {
            return target == sum ? 1 : 0;
        }
        if(dp[cur][total + sum] != -1) return dp[cur][total + sum];
        int ways1 = solve(nums, target, cur+1, sum + nums[cur], dp);
        int ways2 = solve(nums, target, cur+1, sum - nums[cur], dp);
        return dp[cur][total + sum] = ways1 + ways2;
    }
}